src/methods/widen.js

```diff
@@ -0,0 +1,309 @@
+    let [result2, , bindings2, properties2, createdObj2] = e2;
+
+    let result = this.widenResults(realm, result1, result2);
+    let bindings = this.widenBindings(realm, bindings1, bindings2);
+    let properties = this.widenPropertyBindings(realm, properties1, properties2, createdObj1, createdObj2);
+    let createdObjects = new Set();
+    let generator = new Generator(realm); // code subject to widening will be generated somewhere else
+    return [result, generator, bindings, properties, createdObjects];
+  }
+
+  widenResults(realm: Realm, result1: EvaluationResult, result2: EvaluationResult): PossiblyNormalCompletion | Value {
+    invariant(!(result1 instanceof Reference || result2 instanceof Reference), "loop bodies should not result in refs");
+    invariant(
+      !(result1 instanceof AbruptCompletion || result2 instanceof AbruptCompletion),
+      "if a loop iteration ends abruptly, there is no need for fixed point computation"
+    );
+    if (result1 instanceof Value && result2 instanceof Value) {
+      let val = this.widenValues(realm, result1, result2);
+      invariant(val instanceof Value);
+      return val;
+    }
+    if (result1 instanceof PossiblyNormalCompletion && result2 instanceof PossiblyNormalCompletion) {
+      //todo: #1174 figure out how to deal with loops that have embedded conditional exits
+      // widen join pathConditions
+      // widen normal result and Effects
+      // use abrupt part of result2, depend stability to make this safe. See below.
+      throw new FatalError();
+    }
+    // todo: #1174 figure out what a stable result is and how to check it
+    invariant(false, "widening should happen only after result type has stablized");
+  }
+
+  widenMaps<K, V>(m1: Map<K, void | V>, m2: Map<K, void | V>, widen: (K, void | V, void | V) => V): Map<K, void | V> {
+    let m3: Map<K, void | V> = new Map();
+    m1.forEach((val1, key, map1) => {
+      let val2 = m2.get(key);
+      let val3 = widen(key, val1, val2);
+      m3.set(key, val3);
+    });
+    m2.forEach((val2, key, map2) => {
+      if (!m1.has(key)) {
+        m3.set(key, widen(key, undefined, val2));
+      }
+    });
+    return m3;
+  }
+
+  widenBindings(realm: Realm, m1: Bindings, m2: Bindings): Bindings {
+    let widen = (b: Binding, v1: void | Value, v2: void | Value) => {
+      let result = this.widenValues(realm, v1 || b.value, v2 || b.value);
+      invariant(result instanceof Value);
+      return result;
+    };
+    return this.widenMaps(m1, m2, widen);
+  }
+
+  // Returns an abstract value that includes both v1 and v2 as potential values.
+  widenValues(
+    realm: Realm,
+    v1: void | Value | Array<Value> | Array<{ $Key: void | Value, $Value: void | Value }>,
+    v2: void | Value | Array<Value> | Array<{ $Key: void | Value, $Value: void | Value }>
+  ): Value | Array<Value> | Array<{ $Key: void | Value, $Value: void | Value }> {
+    if (Array.isArray(v1) || Array.isArray(v2)) {
```

Comment:
Tricky control flow. If one is an array, the other must also be or `undefined`? Deserves at least a comment.

Reply:
Fair enough.